compiler/src/dotty/tools/dotc/core/Symbols.scala

```diff
@@ -398,19 +398,34 @@ object Symbols {
 
     /** The last denotation of this symbol */
     private[this] var lastDenot: SymDenotation = _
+    private[this] var checkedPeriod: Period = Nowhere
+
+    private[core] def invalidateDenotCache() = { checkedPeriod = Nowhere }
 
     /** Set the denotation of this symbol */
-    private[core] def denot_=(d: SymDenotation) =
+    private[core] def denot_=(d: SymDenotation) = {
       lastDenot = d
+      checkedPeriod = Nowhere
+    }
 
     /** The current denotation of this symbol */
     final def denot(implicit ctx: Context): SymDenotation = {
-      var denot = lastDenot
-      if (!(denot.validFor contains ctx.period)) {
-        denot = denot.current.asInstanceOf[SymDenotation]
-        lastDenot = denot
-      }
-      denot
+      val lastd = lastDenot
+      if (checkedPeriod == ctx.period) lastd
+      else computeDenot(lastd)
+    }
+
+    private def computeDenot(lastd: SymDenotation)(implicit ctx: Context): SymDenotation = {
+      val now = ctx.period
+      checkedPeriod = now
+      if (lastd.validFor contains now) lastd else recomputeDenot(lastd)
+    }
+
+    /** Overridden in NoSymbol */
```

Comment:
Not sure if it'll help performance, but have you tried doing `if (this eq NoSymbol) NoDenotation else ...` to avoid the override?

Reply:
I have not tired it. I don't think we'd see a difference either way. Would be good to know more about tradeoffs like this, though.